plugins/module_utils/attachment.py

```diff
@@ -0,0 +1,175 @@
+        new_payload = payload.copy()
+        new_payload["file_name"] = get_file_name(file_dict)
+        file_type = get_file_type(file_dict)
+
+        if (
+            "encryption_context" in file_dict
+            and file_dict["encryption_context"] is not None
+        ):
+            new_payload["encryption_context"] = file_dict["encryption_context"]
+
+        data = None
+        if not check_mode:
+            with open(file_dict["path"], "rb") as file_obj:
+                data = file_obj.read()
+        return self.create_record(new_payload, data, check_mode, file_type)
+
+    def upload_records(self, payload, file_dict_list, check_mode):
+        return [
+            self.upload_record(payload, file_dict, check_mode)
+            for file_dict in (file_dict_list or [])
+        ]
+
+    def delete_record(self, record, check_mode, silent=False):
+        if not check_mode:
+            if record is None and silent:
+                return {"changed": False, "msg": "Skipped. Record doesn't exist."}
+            else:
+                try:
+                    self.client.delete(_path(record["sys_id"]))
+                    return {"changed": True}
+                except errors.UnexpectedAPIResponse:
+                    return {"changed": False}
+
+    def delete_attached_records(self, payload, check_mode, silent=False):
+        return [
+            self.delete_record(record, check_mode, silent)
+            for record in self.list_records(payload)
+        ]
+
+    def is_changed(self, payload, file_dict):
+        rec = self.get_record(build_query(payload, file_dict))
+        if rec is not None:
+            return file_hash(file_dict["path"]) != rec["hash"]
+        return True
+
+    def are_changed(self, payload, file_dict_list):
+        if file_dict_list is not None:
+            return [
+                self.is_changed(payload, file_dict)
+                for file_dict in (file_dict_list or [])
+            ]
+        return []
+
+    def update_record(self, payload, file_dict, check_mode=False):
+        if self.is_changed(payload, file_dict):
+            self.delete_record(
+                self.get_record(build_query(payload, file_dict)), check_mode, True
+            )
+            return dict(
+                {
+                    "changed": True,
+                    "msg": "Changes detected, hash doesn't match remote. Remote updated.",
+                },
+                **self.upload_record(payload, file_dict, check_mode)
+            )
+        else:
+            return dict(
+                {"changed": False, "msg": "Skipped. Hash matches remote."},
+                **self.get_record(build_query(payload, file_dict))
+            )
+
+    def update_records(self, payload, file_dict_list, check_mode=False):
+        return [
+            self.update_record(payload, file_dict, check_mode)
+            for file_dict in (file_dict_list or [])
+        ]
```

Comment:
We need to modify this method because right now, we are doing way too many API requests (if I counted correctly, we do 2n requests if we have n attachments). I propose the following logic that should drop the API requests to k + 1, where k is the number of records that we actually need to update:

 1. Fetch all metadata for attachments, attached to a record we are trying to update.
 2. Iterate `metadata_list`:
     1. If the current item is not already attached, upload it.
     2. If the current item is attached and the hashes do not match, delete and re-upload the file.

The important part here is: we fetch all existing records in a single call.